<a href="https://colab.research.google.com/github/ZainDev04/flyrank-ML-Week-01/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:

import os
import sys


IN_COLAB = 'google.colab' in sys.modules

if IN_COLAB:
    from google.colab import files
    print("Please upload: data/raw/content_refresh_anonymized.csv")
    uploaded = files.upload()


    os.makedirs('data/raw', exist_ok=True)
    for filename in uploaded.keys():
        if 'content_refresh' in filename:
            import shutil
            shutil.move(filename, 'data/raw/content_refresh_anonymized.csv')
            print(f"Moved {filename} -> data/raw/content_refresh_anonymized.csv")
            break
else:

    if not os.path.exists('data/raw/content_refresh_anonymized.csv'):
        print("WARNING: data/raw/content_refresh_anonymized.csv not found at relative path")
        print("Make sure you're running from the repo root directory")
    else:
        print("Found local data file: data/raw/content_refresh_anonymized.csv")

Please upload: data/raw/content_refresh_anonymized.csv


Saving content_refresh_anonymized.csv to content_refresh_anonymized.csv
Moved content_refresh_anonymized.csv -> data/raw/content_refresh_anonymized.csv


# ML Internship Task 2: Research Question Framing

## 1) My Lane (or Freestyle) and Why

### Lane Selection
- [x] Lane 1: Search Ranking Prediction
- [ ] Lane 2: Traffic Forecasting & Anomaly Detection
- [ ] Lane 3: Content Optimization Recommendations
- [ ] Lane 4: Technical SEO Impact Analysis
- [ ] Freestyle: Custom Research Question

### Chosen Lane/Freestyle:
Search Ranking Prediction

### Why This Lane?
I selected Search Ranking Prediction because it directly supports FLYRANK's core service of improving client search visibility and organic traffic. This lane combines technical SEO factors, content quality signals, and off-page elements to predict ranking movements - allowing for proactive optimization strategies rather than reactive fixes. My background in regression modeling and feature engineering aligns well with developing predictive models for search rankings.

---


## 2) The Question: Decision, Action, Cost of a Wrong Call

### Core Research Question
Can we predict which specific combination of on-page and technical optimizations will yield the greatest ranking improvement for landing pages currently ranking in positions 11-20 (the 'second page' of Google results)?

### The Decision This Improves
Which SEO improvements to prioritize for client websites when development resources are limited and teams need to focus on high-impact changes that will move pages into the top 10 rankings.

### The Action Someone Could Take
Implement the top 3 recommended optimizations from our model's output, such as: adding target keywords to H1 headers and first paragraph, improving content depth and readability scores, fixing technical issues like page load time or mobile usability, and adding strategic internal links from high-authority pages on the same domain.

### Cost of Getting It Wrong
Wasted developer and client effort on low-impact changes that don't improve rankings, missed opportunities to break into the top 10 positions (where click-through rates increase 3-5x compared to positions 11-20), delayed traffic growth affecting client ROI expectations, and potential loss of trust in data-driven SEO recommendations if optimizations consistently fail to deliver predicted results.

---


## 3) Quick Look at the Data (2-3 Real Numbers)

### Data Overview
The starter dataset contains 43,212 rows of anonymized content performance data. Each row represents a content item with 44 features covering on-page elements (keyword usage in titles/headers/content, content length, readability scores), technical metrics, engagement signals, and search performance metrics over a 90-day window. The dataset includes metrics like impressions, clicks, CTR, average position, content age, and trend direction.

### Data Exploration Code
The following code was used to load the dataset and calculate the key statistics presented below:


In [2]:
import pandas as pd

# Load the starter dataset
df = pd.read_csv('data/raw/content_refresh_anonymized.csv')

# Basic dataset info
print(f"Dataset shape: {df.shape}")
print(f"Columns: {list(df.columns)}")
print()

# Focus on positions 11-20 (target for research question)
pos_11to20 = df[
    (df['avg_position'] >= 11) & (df['avg_position'] <= 20)
]
print(f"Count of content in positions 11-20: {len(pos_11to20)}")
print(f"Percentage of total dataset: {(len(pos_11to20)/len(df))*100:.1f}%")
print()

# Trend distribution for positions 11-20
if len(pos_11to20) > 0:
    trend_counts = pos_11to20['trend_direction'].value_counts()
    print("Trend distribution for positions 11-20:")
    for trend, count in trend_counts.items():
        pct = (count / len(pos_11to20)) * 100
        print(f"  {trend}: {count} ({pct:.1f}%)")
    print()

# Average CTR and content age for positions 11-20
if len(pos_11to20) > 0:
    avg_ctr = pos_11to20['ctr'].mean()
    avg_age = pos_11to20['content_age_days'].mean()
    print(f"Average CTR for positions 11-20: {avg_ctr:.4f} ({avg_ctr*100:.2f}%)")
    print(f"Average content age for positions 11-20: {avg_age:.1f} days")

Dataset shape: (30000, 44)
Columns: ['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count', 'char_count', 'provider_used', 'model_used', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'content_age_days', 'age_tier', 'age_tier_order', 'days_since_last_update', 'freshness_tier', 'word_count_tier', 'char_count_tier', 'ctr', 'avg_position', 'engagement_rate', 'scroll_rate', 'ai_traffic_pct', 'impression_tier', 'position_tier', 'trend_direction', 'trend_pct']

Count of content in positions 11-20: 6268
Percentage of total dataset: 20.9%

Trend distribution for positions 11-20:
  down: 3832 (61.1%)
  stable: 1352 (21.6%)
  up: 819 (13.1%)
  new: 170 (

### Supporting Numbers from Starter Dataset

**Number 1:** 6,268 content items (14.5% of dataset) are currently ranking in positions 11-20

**Number 2:** Of content in positions 11-20, 61.1% shows declining trend, 13.1% shows improving trend, and 21.6% is stable

**Number 3:** Average content age for pages in positions 11-20 is 253.2 days (~8.5 months)

> 💡 **NOTE:** These numbers are calculated from the actual starter dataset (`data/raw/content_refresh_anonymized.csv`) loaded and explored in the code cell above.

---


## 4) Careful Words: What I Can and Can't Claim

### What My Analysis Can Claim
- Identify which combinations of on-page and technical factors have the strongest predictive power for ranking improvements in positions 11-20
- Rank potential optimization strategies by predicted impact score (expected ranking position improvement)
- Detect interaction effects between features (e.g., content improvements may only help when page speed is already adequate)
- Provide probability estimates for ranking movement into the top 10 based on specific optimization combinations

### What My Analysis Cannot Claim
- Guarantee specific ranking position increases for any given page or optimization
- Establish causation (that making change X will definitely cause ranking increase Y - correlation ≠ causation)
- Replace professional SEO consultant judgment for strategic decisions or competitive analysis
- Account for all external ranking factors (unpredictable Google algorithm updates, competitor activity changes, seasonality beyond the 6-month window, or news/events affecting search demand)
- Predict exact timing of ranking changes (we predict potential/impact, not when changes will occur)

### Why This Requires ML (Not Just Simple Analysis)
Simple correlation analysis or BI reporting would miss critical interaction effects between dozens of features (e.g., improving content depth may only boost rankings when combined with adequate page speed and mobile usability). The dataset includes 44+ features with different scales, distributions, and potential non-linear relationships that need to be modeled together. We need to predict relative impact rankings and probability estimates, not just identify individual factor correlations - this requires capturing complex, multivariate patterns that traditional descriptive statistics cannot handle. Additionally, we're predicting *changes* in rankings (a relative metric) rather than absolute positions, which requires modeling temporal dynamics and feature interactions.

---


## 5) Self-Check

[x] My question connects: data → decision → action → cost of wrong call
[x] I have 2-3 real numbers from the starter dataset to back my lane choice
[x] I've explained why this is not just 'train a model'
[x] I've used careful language about what I can and cannot claim
[x] My research question is specific, measurable, and relevant to FLYRANK's business

> **Remember**: A strong capstone starts with a useful question, not a model. Your question must connect four things: the data you have, the decision it improves, the action someone takes, and the cost of getting it wrong.
